plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java

```diff
@@ -158,80 +123,56 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
-				FileChangeType changeType = fileEvent.getType();
-
-				if (changeType == FileChangeType.Deleted) {
-					deletedFiles.add(uri);
-				} else {
-					dirtyFiles.add(uri);
-				}
+			FileChangeType changeType = fileEvent.getType();
+			if (changeType == FileChangeType.Deleted) {
+				deletedFiles.add(uri);
+			} else {
+				dirtyFiles.add(uri);
 			}
 		}
 		if (!dirtyFiles.isEmpty() || !deletedFiles.isEmpty()) {
-			runBuildable("didChangeWatchedFiles", () -> buildManager.didChangeFiles(dirtyFiles, deletedFiles));
+			asyncRunBuildTask("didChangeWatchedFiles",
+					() -> buildManager.incrementalBuildTask(dirtyFiles, deletedFiles));
 		}
 	}
 
-	/**
-	 * Evaluate the parameters and deduce the respective build command.
+	// TODO accept a parameter `boolean cancelPrevious` and overload most of the other methods
+	/*
+	 * Use case: We don't want to cancel the clean when we run a reinitWorkspace afterwards.
 	 */
-	protected XBuildable toBuildable(DidSaveTextDocumentParams params) {
-		return buildManager.didSave(getURI(params.getTextDocument()));
-	}
-
 	/**
-	 * Compute a buildable and run it on the queue used for the builder.
+	 * Obtain a buildable and run it on the queue used for the workspace builder.
 	 *
-	 * @param newBuildable
+	 * @param taskSupplier
 	 *            the factory for the buildable.
-	 * @return the result.
 	 */
-	public CompletableFuture<List<Delta>> runBuildable(String description,
-			Supplier<? extends XBuildable> newBuildable) {
-		return queuedExecutorService.submitAndCancelPrevious(XBuildManager.class, description, cancelIndicator -> {
-			XBuildable buildable = newBuildable.get();
-			return buildable.build(cancelIndicator);
+	public CompletableFuture<?> asyncRunBuildTask(String description,
+			Supplier<? extends BuildTask> taskSupplier) {
+		return queuedExecutorService.submitAndCancelPrevious(XWorkspaceBuilder.class, description, cancelIndicator -> {
+			return taskSupplier.get().build(cancelIndicator);
 		});
 	}
 
 	/**
 	 * Initiate an orderly shutdown.
 	 */
-	public CompletableFuture<Void> shutdown() {
-		return runBuildable("shutdown", () -> {
-			return (cancelIndicator) -> {
-				joinPersister();
-				return null;
-			};
-		}).thenApply(any -> {
-			persister.close();
-			queuedExecutorService.shutdown();
-			return null;
-		});
+	public void shutdown() {
+		join();
+		queuedExecutorService.shutdown();
 	}
 
 	/**
 	 * Block until all submitted background work is done.
 	 */
 	public void join() {
 		queuedExecutorService.join();
-		joinPersister();
-	}
-
-	private void joinPersister() {
-		persister.pendingWrites().join();
 	}
 
 	/**
-	 * Answer true, if the uri is from a source folder in the current workspace.
+	 * Returns the workspace issues known for the given URI.
 	 */
-	protected boolean isSourceFile(URI uri) {
-		return workspaceManager.isSourceFile(uri);
+	public ImmutableList<? extends LSPIssue> getValidationIssues(URI uri) {
+		return workspaceManager.getValidationIssues(uri);
```

Comment:
Maybe I'm overlooking something, but this method is invoked from arbitrary threads, so it should be thread-safe, right? Then, this also applies to all methods in the internal builder code that are directly or indirectly invoked from here, for example `XWorkspaceManager#getProjectBuilder(String)`, `ProjectBuilder#getValidationIssues(URI)`, etc. To avoid multi-threading to "bleed into" the builder code in this way was the purpose of the concurrent issue registry. But maybe there's some other way to do this?

Reply:
In a subsequent change, we'll encapsulate the state of the project in immutable data objects that will be updated after the project build in an atomic step (think IndexState on steroids). For now I'll add synchronization.